### ISSUE TITLE
Limiting number of concurrent processes and their resources

### ISSUE THREAD

**Issue by infokiller** [glacambre/firenvim#241]
The [security doc](https://github.com/glacambre/firenvim/blob/master/SECURITY.md#malicious-page) mentions that a malicious page can cause resource exhaustion, and suggests limiting the number of concurrent processes. I wanted to start a discussion about what's needed to implement this limitation.

Ideally, not only the number of processes will be limited, but also the total resources they consume, though this may introduce too much complexity. I'm mostly using Linux, and for my specific needs I can probably use cgroups to limit the combined resources of the firenvim, but I'm not sure about other OS.

Another thing that should be limited is the rate of creating new instances in a page. Otherwise, if only the number of concurrent instances is limited, a malicious page can start and exit instances in a loop.

**Comment by glacambre** [comment]
I think limiting the total amount of resources consumed by firenvim (well, neovim would actually be the one consuming the resources) would be far too complex to implement. We could document ways to do that with external tools but I don't want firenvim to deal with it itself.

I don't think limiting the number of concurrent instances in a tab/page would be useful, I would prefer a single global number of instances. In my opinion, what we want is to prevent a malicious actor from rendering your computer unusable - if you're just trying to open firenvim in a new tab and you get an error telling you that there are too many firenvim instances open, you can just close a few tabs until you find the culprit.

Implementing a limit on the number of neovim processes shouldn't be too hard. Here's what I envision:
- Create a new key named "globalSettings" in the firenvim_config object
- This globalSettings object would contain a key named e.g. "maxInstances" (we should find a better name), the value of which is the number of concurrent neovim instances firenvim is allowed to have.
- This maxInstances setting should have a [default value](https://github.com/glacambre/firenvim/blob/30d0dc8c00e89259369471a27f48ab00af2ad531/src/background.ts#L154) - perhaps 10?
- Then, we could just increment a number when [a new instance is created](https://github.com/glacambre/firenvim/blob/30d0dc8c00e89259369471a27f48ab00af2ad531/src/background.ts#L180) and decrement it when the [onDisconnect](https://developer.mozilla.org/en-US/docs/Mozilla/Add-ons/WebExtensions/API/runtime/Port) event is triggered.
- If the maximal number of allowed instances is created, [createNewInstance](https://github.com/glacambre/firenvim/blob/30d0dc8c00e89259369471a27f48ab00af2ad531/src/background.ts#L179) should just return a rejected promise, set the [global error message](https://github.com/glacambre/firenvim/blob/master/src/background.ts#L62) and [update the browserAction icon](https://github.com/glacambre/firenvim/blob/30d0dc8c00e89259369471a27f48ab00af2ad531/src/background.ts#L41).
Of course this would need at least one test, which would [set the setting](https://github.com/glacambre/firenvim/blob/30d0dc8c00e89259369471a27f48ab00af2ad531/tests/_common.ts#L365-L372), [load a custom page](https://github.com/glacambre/firenvim/blob/30d0dc8c00e89259369471a27f48ab00af2ad531/tests/_common.ts#L49) containing N+1 textareas, [select them all](https://github.com/glacambre/firenvim/blob/30d0dc8c00e89259369471a27f48ab00af2ad531/tests/_common.ts#L375-L380) and check that the N+1th [wasn't created](https://github.com/glacambre/firenvim/blob/30d0dc8c00e89259369471a27f48ab00af2ad531/tests/_common.ts#L349-L357).

**Comment by infokiller** [comment]
> I think limiting the total amount of resources consumed by firenvim (well, neovim would actually be the one consuming the resources) would be far too complex to implement. We could document ways to do that with external tools but I don't want firenvim to deal with it itself.

That sounds like a good compromise.

> 
> I don't think limiting the number of concurrent instances in a tab/page would be useful, I would prefer a single global number of instances. In my opinion, what we want is to prevent a malicious actor from rendering your computer unusable - if you're just trying to open firenvim in a new tab and you get an error telling you that there are too many firenvim instances open, you can just close a few tabs until you find the culprit.

Both options (limiting per page and total) seem reasonable to me.
The pro of limiting per page is that it makes it clearer to the user what went wrong when the limit is reached, as they can have many pages open. I frequently have 200+ tabs open across multiple windows, and finding tabs with open firenvim instances may be time consuming.

I related though I have (and please let me know if this should go in a separate issue) is that of "garbage collection". What's the current story for instances that were opened a long time ago?
It may be useful to have some system to periodically quit all instances that are idle for a long time.

> 
> Implementing a limit on the number of neovim processes shouldn't be too hard. Here's what I envision:
> 
> * Create a new key named "globalSettings" in the firenvim_config object
> * This globalSettings object would contain a key named e.g. "maxInstances" (we should find a better name), the value of which is the number of concurrent neovim instances firenvim is allowed to have.
> * This maxInstances setting should have a [default value](https://github.com/glacambre/firenvim/blob/30d0dc8c00e89259369471a27f48ab00af2ad531/src/background.ts#L154) - perhaps 10?
> * Then, we could just increment a number when [a new instance is created](https://github.com/glacambre/firenvim/blob/30d0dc8c00e89259369471a27f48ab00af2ad531/src/background.ts#L180) and decrement it when the [onDisconnect](https://developer.mozilla.org/en-US/docs/Mozilla/Add-ons/WebExtensions/API/runtime/Port) event is triggered.
> * If the maximal number of allowed instances is created, [createNewInstance](https://github.com/glacambre/firenvim/blob/30d0dc8c00e89259369471a27f48ab00af2ad531/src/background.ts#L179) should just return a rejected promise, set the [global error message](https://github.com/glacambre/firenvim/blob/master/src/background.ts#L62) and [update the browserAction icon](https://github.com/glacambre/firenvim/blob/30d0dc8c00e89259369471a27f48ab00af2ad531/src/background.ts#L41).
> 
> Of course this would need at least one test, which would [set the setting](https://github.com/glacambre/firenvim/blob/30d0dc8c00e89259369471a27f48ab00af2ad531/tests/_common.ts#L365-L372), [load a custom page](https://github.com/glacambre/firenvim/blob/30d0dc8c00e89259369471a27f48ab00af2ad531/tests/_common.ts#L49) containing N+1 textareas, [select them all](https://github.com/glacambre/firenvim/blob/30d0dc8c00e89259369471a27f48ab00af2ad531/tests/_common.ts#L375-L380) and check that the N+1th [wasn't created](https://github.com/glacambre/firenvim/blob/30d0dc8c00e89259369471a27f48ab00af2ad531/tests/_common.ts#L349-L357). It would need to run in both [firefox](https://github.com/glacambre/firenvim/blob/30d0dc8c00e89259369471a27f48ab00af2ad531/tests/firefox.ts#L69) and [chrome](https://github.com/glacambre/firenvim/blob/30d0dc8c00e89259369471a27f48ab00af2ad531/tests/chrome.ts#L68).

Thanks for the detailed instructions, they're very useful!
I'll try to carve out some time to work on this (but I'll need to familiarize myself with some of the tech used prior to that).

**Comment by glacambre** [comment]
>The pro of limiting per page is that it makes it clearer to the user what went wrong when the limit is reached, as they can have many pages open. I frequently have 200+ tabs open across multiple windows, and finding tabs with open firenvim instances may be time consuming.

A per-page setting wouldn't prevent an attacker from causing resource exhaustion as they could just open multiple frames with different domains.

>I related though I have (and please let me know if this should go in a separate issue) is that of "garbage collection". What's the current story for instances that were opened a long time ago?
>It may be useful to have some system to periodically quit all instances that are idle for a long time.

This can already be implemented from Neovim, create a timer (see `:h timer`) that gets reset on TextChanged events and that kills the current neovim instance on completion.

**Comment by infokiller** [comment]
> > The pro of limiting per page is that it makes it clearer to the user what went wrong when the limit is reached, as they can have many pages open. I frequently have 200+ tabs open across multiple windows, and finding tabs with open firenvim instances may be time consuming.
> 
> A per-page setting wouldn't prevent an attacker from causing resource exhaustion as they could just open multiple frames with different domains.

How will the attacker be able to open multiple frames with different domains from a single tab?
I'm assuming the attacker can't open a number of tabs without being blocked by the browser.
> 
> > I related though I have (and please let me know if this should go in a separate issue) is that of "garbage collection". What's the current story for instances that were opened a long time ago?
> > It may be useful to have some system to periodically quit all instances that are idle for a long time.
> 
> This can already be implemented from Neovim, create a timer (see `:h timer`) that gets reset on TextChanged events and that kills the current neovim instance on completion.

This should work, thanks! I'll try it.

**Comment by glacambre** [comment]
>How will the attacker be able to open multiple frames with different domains from a single tab?

`document.body.appendChild(document.createElement("iframe"))`

**Comment by infokiller** [comment]
So it's possible to open an arbitrary number of iframes from a single document, and then, say, use javascript to focus on different text areas from different iframes?

**Comment by glacambre** [comment]
Yep.

**Comment by glacambre** [comment]
Now that I've made progress on #197 I think it'll be enough to prevent PID/Port exhaustion, so I won't implement any other mechanism to close this issue. Closing in favor of #197.
